nexus/src/app/instance.rs

```diff
@@ -287,39 +305,107 @@ impl super::Nexus {
+    /// successfully reflected into CRDB. Returns `Err` with an appropriate
+    /// error otherwise.
+    ///
+    /// # Panics
+    ///
+    /// Asserts that `db_instance` has no migration ID or destination Propolis
+    /// ID set.
+    pub async fn instance_set_migration_ids(
         &self,
-        _opctx: &OpContext,
-        _instance_id: Uuid,
-        _migration_id: Uuid,
-        _dst_propolis_id: Uuid,
+        opctx: &OpContext,
+        instance_id: Uuid,
+        db_instance: &db::model::Instance,
+        migration_params: InstanceMigrationSourceParams,
     ) -> UpdateResult<db::model::Instance> {
-        todo!("Migration endpoint not yet implemented in sled agent");
+        assert!(db_instance.runtime().migration_id.is_none());
+        assert!(db_instance.runtime().dst_propolis_id.is_none());
 
-        /*
-        let (.., authz_instance, db_instance) =
-            LookupPath::new(opctx, &self.db_datastore)
-                .instance_id(instance_id)
-                .fetch()
-                .await
-                .unwrap();
-        let requested = InstanceRuntimeStateRequested {
-            run_state: InstanceStateRequested::Migrating,
-            migration_params: Some(InstanceRuntimeStateMigrateParams {
-                migration_id,
-                dst_propolis_id,
-            }),
-        };
-        self.instance_set_runtime(
-            opctx,
-            &authz_instance,
-            &db_instance,
-            requested,
-        )
-        .await?;
-        self.db_datastore.instance_refetch(opctx, &authz_instance).await
-        */
+        let (.., authz_instance) = LookupPath::new(opctx, &self.db_datastore)
+            .instance_id(instance_id)
+            .lookup_for(authz::Action::Modify)
+            .await
+            .unwrap();
+
+        let sa = self.instance_sled(&db_instance).await?;
+        let instance_put_result = sa
+            .instance_put_migration_ids(
+                &instance_id,
+                &InstancePutMigrationIdsBody {
+                    old_runtime: db_instance.runtime().clone().into(),
+                    migration_params: Some(migration_params),
+                },
+            )
+            .await
+            .map(|res| Some(res.into_inner()));
+
+        // Write the updated instance runtime state back to CRDB. If this
+        // outright fails, this operation fails. If the operation nominally
+        // succeeds but nothing was updated, this action is outdated and the
+        // caller should not proceed with migration.
+        let updated = self
+            .handle_instance_put_result(&db_instance, instance_put_result)
+            .await?;
+
+        if updated {
+            Ok(self
+                .db_datastore
+                .instance_refetch(opctx, &authz_instance)
+                .await?)
+        } else {
+            Err(Error::internal_error(
+                "instance's Propolis generation is out of date",
+            ))
```

Comment:
668502f87 switches this from Internal Error to Unavailable. I thought about 409 Conflict, but (a) there's no `Error` variant for that (not that I couldn't create one, I suppose), and (b) that doesn't seem quite correct semantically since this could happen for reasons totally unrelated to any prior command the caller issued (e.g. it's not necessarily the caller's fault if the instance stopped just as the saga tried to set migration IDs on it).

I also tweaked the error message to remove the reference to "Propolis"--I'm not sure exactly how this error bubbles up but it seemed the safest course not to refer to any inside baseball here.